walletrpc: unify coin selection, allow coin selection for PSBT with pre-defined inputs

guggero opened the pull request:
Depends on https://github.com/btcsuite/btcwallet/pull/900.

This PR first adds the coin selection strategy to the channel funding coin selection algorithm.
Then we prepare the coin selection algorithm in the `chanfunding` package to be re-used by other packages by generalizing some of the assumptions made regarding weights and change output sizes.

Finally we add a new, third, mode for funding a PSBT where we allow inputs to be pre-defined while still performing coin selection (previously users could only either specify no inputs at all to trigger coin selection, or, if inputs were specified, no new ones would be selected and only fee estimation would be performed).

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
    - Added `CoinSelectionStrategy` to wallet configuration options.

- **Enhancements**
    - Enhanced address properties to include derivation path index.

- **Refactor**
    - Updated coin selection logic and types used in `chanfunding`.

- **Bug Fixes**
    - Corrected the weight estimation of PSBT inputs.

- **Documentation**
    - Updated release notes to reflect new features and improvements.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/lightningnetwork/lnd/8378)
<!-- Reviewable:end -->

Roasbeef commented on lnrpc/walletrpc/walletkit_server.go:
```diff
@@ -1378,6 +1420,180 @@ func (w *WalletKit) fundPsbtInternalWallet(account string,
+		changeOutput := packet.UnsignedTx.TxOut[changeIndex]
+		changeScriptLen = len(changeOutput.PkScript)
+
+	case changeType == ChangeAddressType_CHANGE_ADDRESS_TYPE_P2TR:
```
Will P2TR be the default change type here?

guggero replied:
I would say no. Because the default for the normal `FundPsbt` is P2WKH (which was kept to not break backward compatibility with earlier RPC users, see more context in discussion here: https://github.com/lightningnetwork/lnd/pull/7348).
And since the change type is on the parent message, I don't think that using the new third funding option should introduce a different behavior of a top-level parameter.
But since we can specify the change type to be P2TR, I don't think the default is a big deal.

Roasbeef commented on lnwallet/chanfunding/coin_select.go:
```diff
@@ -87,10 +87,11 @@ func selectInputs(amt btcutil.Amount, coins []Coin,
+	changeScriptLen int) (btcutil.Amount, btcutil.Amount, error) {
 
-	var weightEstimate input.TxWeightEstimator
+	weightEstimate := existingWeight
```
👍

guggero commented on config_builder.go:
```diff
@@ -686,15 +686,16 @@ func (d *DefaultWalletImpl) BuildChainControl(
+		ChainIO:               walletController,
+		DefaultConstraints:    partialChainControl.ChannelConstraints,
+		NetParams:             *walletConfig.NetParams,
+		CoinSelectionStrategy: walletConfig.CoinSelectionStrategy,
```
Yeah, I think that makes sense. Though it feels like that would be a different PR, as we'd probably also want to add such new parameters to the existing RPCs that currently use the config-level strategy (`FundPsbt`, `EstimateFee`, `SendCoins`, `SendOutputs`).

guggero replied:
https://github.com/lightningnetwork/lnd/issues/8498

guggero commented on lnrpc/walletrpc/walletkit.proto:
```diff
@@ -1144,6 +1144,19 @@ message FundPsbtRequest {
+        already specified in the PSBT must already be locked (if they belong to
+        this node), only newly added inputs will be locked by this RPC.
+        */
+        PsbtCoinSelect coin_select = 9;
```
Yes, that's one instance where that will become useful! Unfortunately we'll only be able to replace the code once we have `lnd v0.18` out and `tapd` having hard dependency requirement on it.

guggero commented on input/size.go:
```diff
@@ -1013,6 +1020,17 @@ func (twe *TxWeightEstimator) AddP2SHOutput() *TxWeightEstimator {
+// AddOutputWithScriptLength estimates the weight of an output based on the
+// pkScript length.
+func (twe *TxWeightEstimator) AddOutputWithScriptLength(
+	scriptLen int) *TxWeightEstimator {
```
This probably won't make it into the final PR. This is needed for adding the change output weight. And because we want to avoid deriving a change address before we actually need one, I just pass around the expected change script size. But I'll probably replace that with an actual change address type and then use that to estimate the weight.

guggero commented:
Fixed up a bunch of things and added unit and integration tests.
Ready for full review now, thought I might still add more test cases to the PR later on.

dstadulis commented:
@coderabbitai review

guggero commented on lnwallet/wallet.go:
```diff
@@ -2526,7 +2530,7 @@ func NewCoinSource(w *LightningWallet) *CoinSource {
 // minConfs and maxConfs number of confirmations.
 func (c *CoinSource) ListCoins(minConfs int32,
-	maxConfs int32) ([]chanfunding.Coin, error) {
+	maxConfs int32) ([]wallet.Coin, error) {
```
Do you realize your proposed diff is actually empty? I think you're confused.

guggero commented on lnrpc/walletrpc/walletkit_server.go:
```diff
@@ -1349,12 +1451,266 @@
+	changeType chanfunding.ChangeAddressType,
+	feeRate chainfee.SatPerKWeight) (*FundPsbtResponse, error) {
+
+	// Before we select anything, we need to calculate the input, output and
+	// current weight amounts. While doing that we also ensure the PSBT has
+	// all the required information we require at this step.
+	var (
+		inputSum, outputSum btcutil.Amount
+		estimator           input.TxWeightEstimator
+		buf                 bytes.Buffer
+	)
+	for i := range packet.Inputs {
+		in := packet.Inputs[i]
+
+		err := btcwallet.EstimateInputWeight(&in, &estimator)
+		if err != nil {
+			return nil, fmt.Errorf("error estimating input "+
+				"weight: %w", err)
+		}
+
+		inputSum += btcutil.Amount(in.WitnessUtxo.Value)
+	}
+	for i := range packet.UnsignedTx.TxOut {
+		out := packet.UnsignedTx.TxOut[i]
+
+		estimator.AddOutput(out.PkScript)
+		outputSum += btcutil.Amount(out.Value)
+	}
+
+	// The amount we want to fund is the total output sum plus the current
+	// fee estimate. Since we pass the estimator of the current transaction
+	// into the coin selection algorithm, we don't need to subtract the fees
+	// here.
+	fundingAmount := outputSum - inputSum
```
We do it a couple of lines further down.

guggero commented:
@coderabittai pause

ellemouton reviewed (commented):
Very nice! Excellent tests & commit structure! Learned a bunch while reviewing this. 

I left 1 main question (mainly an understanding thing on my part I think) but this looks pretty much g2g 🚀

guggero commented on itest/lnd_channel_funding_utxo_selection_test.go:
```diff
@@ -181,9 +179,9 @@ func testChannelUtxoSelection(ht *lntest.HarnessTest) {
 			localAmt: btcutil.Amount(300_000) -
-				fundingFee(3, true),
+				fundingFee(2, true),
 			expectedBalance: btcutil.Amount(300_000) -
-				fundingFee(3, true),
+				fundingFee(2, true),
```
Yes, this is a slight change in the strategy. Thought perhaps only in the unit tests. I assume that the coins already have some valid order when being returned from the wallet, but not sure what order that is.

There is a global config flag `--coin-selection-strategy=[largest|random]` that now applies to channel funding as well with this PR (previously was only used for `SendCoins` and a few other RPCs).

guggero commented on lnwallet/chanfunding/coin_select.go:
```diff
@@ -185,53 +185,93 @@ func CoinSelect(feeRate chainfee.SatPerKWeight, amt, dustLimit btcutil.Amount,
+	// change output, we instead just let the dust amount go to
+	// fees. Unless we want the change to go to an existing output,
+	// in that case we can increase that output value by any amount.
+	if changeAmt < dustLimit && changeType != ExistingChangeAddress {
```
Right. Though this is refactored later so it ends up with one level less indentation, that's why the linter doesn't complain. Okay if I leave it that way? Otherwise the refactor commit diff is a bit bigger.

guggero commented on lnwallet/btcwallet/psbt.go:
```diff
@@ -352,6 +369,72 @@ func validateSigningMethod(in *psbt.PInput) (input.SignMethod, error) {
+
+	pkScript := in.WitnessUtxo.PkScript
+	switch {
+	case txscript.IsPayToScriptHash(pkScript):
+		w.AddNestedP2WKHInput()
```
Yes, theoretically. But because the `lnd` wallet only supports NP2WKH and no bare/standard P2SH it will never select or be able to sign for such an input. And we wouldn't know the length of the witness anyway, so we couldn't make a better guess. But I think we can just assume nobody is going to use a P2SH input foreign to `lnd` in this RPC? And if they do, the fee estimation will just be off...

guggero commented on lnrpc/walletrpc/walletkit.proto:
```diff
@@ -1301,6 +1322,38 @@ message TxTemplate {
+        /*
+        Use the existing output within the template PSBT with the specified
+        index ax the change output. Any leftover change will be added to the
+        already specified amount of that output. To add a new change output to
```
When funding the PSBT no input should be signed yet. You are correct about some input types requiring a witness, though the idea was that those witnesses would be dummy witness data just for the fee estimation.
But as discussed offline that is probably more confusing than anything else, so I'm going to remove that again.

guggero commented on itest/lnd_psbt_test.go:
```diff
@@ -1032,6 +1039,324 @@ func runFundAndSignPsbt(ht *lntest.HarnessTest, alice *node.HarnessNode) {
+		PkScript: addressToPkScript(ht, bobAddr.Address),
+		Value:    bobUtxo.AmountSat - sendAmount,
+	}, &wire.TxOut{
+		// Amount to be sent to Alice, but we'll also send her change
+		// here.
```
There shouldn't be a signature yet. Discussed offline.

guggero commented on itest/lnd_psbt_test.go:
```diff
@@ -1032,6 +1039,324 @@ func runFundAndSignPsbt(ht *lntest.HarnessTest, alice *node.HarnessNode) {
+	txFee, err := fundedPacket.GetTxFee()
+	require.NoError(ht, err)
+
+	// We now let Bob sign the transaction.
```
Discussed offline.

guggero commented on lnrpc/walletrpc/walletkit_server.go:
```diff
@@ -1373,6 +1448,192 @@ func (w *WalletKit) fundPsbtInternalWallet(account string,
+		// We shouldn't get into this branch if the input sum isn't
+		// enough to pay for the current package without a change
+		// output. So this should never be non-zero.
+		if needMore != 0 {
```
Yes, you are correct, that doesn't necessarily follow. But if there is enough change to warrant a change output, then a change output is added. If there is not enough change for a change output, then none will be added. And if we are somewhere in the middle (e.g. 5 satoshis of change), then that just goes to fees.

Or to to put it another way: In `CalculateChangeAmount`, the only case where `needMore` would be returned as non-zero is when `overShootAmt < requiredFeeNoChange`, which is basically excluded by the `inputSum >= outputSum+packetFeeNoChange` condition above.

This should also be covered by `TestCalculateChangeAmount`. But coin selection is a really hard thing to do, so please let me know if there are any residual doubts and I'll take another look.

guggero commented:
Rebased and added a new `lncli` subcommand: `lncli wallet psbt fundtemplate` that hooks into the new funding option (was easier to create a new command than having multiple flags that are only valid in certain combinations.

ellemouton commented on cmd/lncli/walletrpc_active.go:
```diff
@@ -726,6 +734,339 @@ type fundPsbtResponse struct {
+	The 'outputs' flag decodes addresses and the amount to send respectively
+	in the following JSON format:
+
+	    --outputs='["ExampleAddr:NumCoinsInSatoshis", "SecondAddr:Sats"]'
```
took me a while to realise that this command expects the format of the outputs using `"["` brackets instead of `"{"` brackets like the `fund` command 😅 

As is done for the `fund` command, should we also just have an example of what the input list looks like?

guggero replied:
Yeah, I know... I'd like to change the format of the `fund` command too, but not sure if that would break some existing shell scripts perhaps?

Added the example for the inputs though.

ellemouton commented on cmd/lncli/walletrpc_active.go:
```diff
@@ -726,6 +734,339 @@ type fundPsbtResponse struct {
+	array is used rather than a map.
+
+	After verifying and possibly adding new inputs, all input UTXOs added by
+	the command are locked with an internal app ID. Inputs already present
+	in the template are NOT locked.
```
why not? Is there a way for a user to manually go lock the output then instead? 

cause it seems dangerous to start a psbt round with an input that LND might sweep out from under you for something else if we dont lock them

ellemouton replied:
also - this is different from the behaviour of the `fund` command.

ellemouton replied:
One other thing I noticed is: For the `fund` command, if I specify an input - it get's locked. Then if i try use that same input in another `fund` round - it fails due to `input 0 not found in list of non-locked UTXO` which seems good as this protects me. 

But with `fundtemplate` - im able to specify inputs of mine that are already locked. This seems dangerous. 

I get that we cant lock all the inputs since we dont necessarily own all the inputs, but maybe we should lock the ones that we do own?

guggero replied:
So the locking here is mostly to avoid an input specified in the template to be selected again by the coin selection algorithm. And also to make sure no _other_ application snatches the input away from us while we're using it.

So to your first question "why not" in the CLI description: Because we want the inputs specified in the template to _already_ be locked by the user. We didn't enforce that previously, but since that's dangerous, I now added that check.

> But with `fundtemplate` - im able to specify inputs of mine that are already locked. This seems dangerous.

Hmm, I'm not sure there's a lot we can do about that. We want the user to go ahead and select an input, manually lock it and then call `fundtemplate`. But we cannot really determine whether the user locked them manually or they are locked from a previous attempt. We could ask the user to provide the lock ID they used to lock them, to make sure they did it themselves. But then they can still just provide the lock ID of another process, as you can read the lock ID from the output of `lncli wallet listleases`.
So I think users concerned about using this RPC correctly would hopefully read the instructions and do the locking properly?
Or do you have an idea for how to make things even more explicit?

saubyk replied:
> We could ask the user to provide the lock ID they used to lock them, to make sure they did it themselves

Jumping on a moving train here, so apologies if my comments are completely off-balance. But as a general UX rule, we should try to minimize the inputs user must provide.

ellemouton reviewed (commented):
tACK  🔥 

Had loads of fun testing this out!! Great work 🚀 

One last question around why we dont lock any already specified inputs.

guggero commented on cmd/lncli/walletrpc_active.go:
```diff
@@ -726,6 +734,339 @@ type fundPsbtResponse struct {
 	Locks             []*utxoLease `json:"locks"`
 }
 
+var fundTemplatePsbtCommand = cli.Command{
```
Done. Also mentioned the different format of `--outputs` (and the reason for it).

ellemouton commented on lnrpc/walletrpc/walletkit_server.go:
```diff
@@ -1385,6 +1463,242 @@ func (w *WalletKit) fundPsbtInternalWallet(account string,
+	// either not belong to this lnd at all or to be already locked through
+	// a manual lock call by the user. Either way, they should not appear in
+	// the list of unspent outputs.
+	err := w.assertNotAvailable(packet.UnsignedTx.TxIn, minConfs, account)
```
👍

ellemouton reviewed (approved):
Awesome work 🧩

lightninglabs-deploy commented:
@roasbeef: review reminder
@morehouse: review reminder
@guggero, remember to re-request review from reviewers when ready

Roasbeef commented on itest/lnd_psbt_test.go:
```diff
@@ -1032,6 +1039,324 @@ func runFundAndSignPsbt(ht *lntest.HarnessTest, alice *node.HarnessNode) {
+	for _, account := range addresses.AccountWithAddresses {
+		for _, address := range account.Addresses {
+			if address.Address == addr {
+				path, err := lntest.ParseDerivationPath(
```
Ah ok, I see here that usage is typically done alongside the corresponding account information. Including the full path as a field in the addr can potentially simplify usage for users/devs and also this section (no need for the account look up and match loop).

guggero replied:
Was able to drastically simplify this by adding the full derivation path and public key to each address.

Roasbeef reviewed (approved):
LGTM 👑

Roasbeef commented:
Needs a rebase!

Also will double check that comment I made fee estimation being wrong for P2TR channels, as it is (?) assuming P2WSH is used for the funding output.

guggero commented on lnwallet/chanfunding/coin_select.go:
```diff
@@ -92,17 +106,23 @@ func calculateFees(utxos []wallet.Coin,
 	}
 
-	// Channel funding multisig output is P2WSH.
-	weightEstimate.AddP2WSHOutput()
```
Yes, through the code you commented below as well. P2TR is the same output size, so we don't need the distinction on that level. See comment below.

guggero commented on lnwallet/chanfunding/wallet_assembler.go:
```diff
@@ -308,6 +308,17 @@ func (w *WalletAssembler) ProvisionChannel(r *Request) (Intent, error) {
+		// the same length). And we currently don't support specifying a
+		// change output type, so we always use P2TR.
+		var fundingOutputWeight input.TxWeightEstimator
+		fundingOutputWeight.AddP2WSHOutput()
```
This is what I tried to explain in the comment right above this code. Basically both have the same size so no distinction is needed.

guggero commented on lnrpc/walletrpc/walletkit.proto:
```diff
@@ -517,6 +517,9 @@ message AddressProperty {
     int64 balance = 3;
+
+    // The index of the address in the derivation path.
+    uint32 index = 4;
```
Yes, I think that makes sense. And the public key as well, which also simplifies a lot of things in some of the TAP tests.

guggero commented:
@Roasbeef I rebased and addressed your comments. The address list now returns the full derivation path and public key for each address.

Roasbeef reviewed (approved):
> Reviewed 3 of 17 files at r1, 17 of 23 files at r2, 1 of 2 files at r3, 1 of 2 files at r4, 12 of 12 files at r5, 7 of 20 files at r6, 15 of 15 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/lightningnetwork/lnd/8378)* status: all files reviewed, 46 unresolved discussions (waiting on @ellemouton, @guggero, @jharveyb, and @saubyk)

___
*[`lnwallet/btcwallet/psbt_test.go` line 502 at r7](https://reviewable.io/reviews/lightningnetwork/lnd/8378#-Nrbs1yUCZEUG3-1jToC:-Nrbs1yVAYRbgEr9J1Df:b-8fx2u7) ([raw file](https://github.com/lightningnetwork/lnd/blob/c307d1eeda3d8da344c65c7fbe647d799c02777c/lnwallet/btcwallet/psbt_test.go#L502)):*
> ```Go
> // TestBip32DerivationFromKeyDesc tests that we can correctly extract a BIP32
> // derivation path from a key descriptor.
> func TestBip32DerivationFromKeyDesc(t *testing.T) {
> ```

:+1: 


<!-- Sent from Reviewable.io -->
